File: 704_binary_search/Solution.py
class Solution:
    def search(self, nums: list, target: int) -> int:
        if not nums: return -1

        l, r = 0, len(nums) - 1
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] == target: return mid
            elif nums[mid] > target: r = mid - 1
            else: l = mid + 1

        return -1


if __name__ == '__main__':
    nums = [-1, 0, 3, 5, 9, 12]
    print(Solution().search(nums, 9))

File: 704_binary_search/Solution.go
package main

import "fmt"

func search(nums []int, target int) int {
    if len(nums) == 0 { return -1 }

    l, r := 0, len(nums) - 1
    for l <= r {
        mid := l + (r - l) / 2
        if (nums[mid] == target) {
            return mid
        } else if (nums[mid] < target) {
            l = mid + 1
        } else {
            r = mid - 1
        }
    }

    return -1
}

func main() {
    nums := []int{-1, 0, 3, 5, 9, 12}
    fmt.Println(search(nums, 9))
}
